File: ios/PebbleAudioKit/Sources/SegmentStore/SegmentSink.swift
import Foundation
import WireProtocol

// Port of `core/transport/.../SegmentSink.kt` — the durable-storage seam between the receiver
// session and the segment store. Declared here (rather than in Receiver) so Receiver can depend
// on SegmentStore without a cycle. This file is the FIXED cross-module contract: change it only
// with a matching change on both sides.

/// One encoded audio frame, addressed by stream sequence and cumulative sample index.
public struct SegmentFrame: Sendable, Equatable {
    public let sequence: UInt32
    public let sampleIndex: UInt64
    public let payload: [UInt8]

    public init(sequence: UInt32, sampleIndex: UInt64, payload: [UInt8]) {
        self.sequence = sequence
        self.sampleIndex = sampleIndex
        self.payload = payload
    }
}

/// Why a gap record exists in a segment.
public enum GapOrigin: Sendable, Equatable {
    /// The watch explicitly reported the gap via STREAM_GAP.
    case watchReported(reasonRaw: Int, watchDropCounter: UInt32)
    /// Synthesized by the receiver after observing a sequence discontinuity in STREAM_DATA.
    case sequenceSkip
}

public struct GapRecord: Sendable, Equatable {
    public let firstMissingSequence: UInt32
    /// 0 = unknown / elapsed-time-only gap.
    public let missingFrameCount: UInt32
    public let firstMissingSampleIndex: UInt64
    public let origin: GapOrigin

    public init(
        firstMissingSequence: UInt32,
        missingFrameCount: UInt32,
        firstMissingSampleIndex: UInt64,
        origin: GapOrigin
    ) {
        self.firstMissingSequence = firstMissingSequence
        self.missingFrameCount = missingFrameCount
        self.firstMissingSampleIndex = firstMissingSampleIndex
        self.origin = origin
    }
}

/// Why a segment was closed.
public enum SegmentCloseReason: Sendable, Equatable {
    /// The watch sent STREAM_STOP.
    case stopped(reasonRaw: Int, finalSequence: UInt32, finalSampleIndex: UInt64)
    /// The BLE link dropped (or the session ended) while the stream was open.
    case interrupted
    /// A new STREAM_START arrived while a segment was still open.
    case superseded
}

/// Diagnostics-only provenance attached to a segment.
public struct SegmentProvenance: Sendable, Equatable {
    public let fwVersionPacked: UInt32
    public let protocolVersion: Int

    public init(fwVersionPacked: UInt32, protocolVersion: Int) {
        self.fwVersionPacked = fwVersionPacked
        self.protocolVersion = protocolVersion
    }
}

/// Durable storage seam consumed by `ReceiverSession`; implemented by SegmentStore.
/// All functions are async and must only return once the data is durable — the session
/// computes checkpoint sequences from what these calls have accepted.
public protocol SegmentSink: Sendable {
    /// Opens a segment for a newly started stream. Any previously open segment was closed.
    func openSegment(start: StreamStart, receivedAtMs: Int64, provenance: SegmentProvenance?) async throws

    /// Appends frames to the open segment; durable on return. Returns the frames actually
    /// persisted — a post-RESUME spool rewind re-sends frames the sink may already hold, and
    /// downstream consumers (live waveform, live transcription) must only see what was new.
    func appendFrames(streamId: UInt32, frames: [SegmentFrame]) async throws -> [SegmentFrame]

    /// Records a gap (watch-reported or synthesized) in the open segment.
    func recordGap(streamId: UInt32, gap: GapRecord) async throws

    /// Closes the open segment. No-op when no segment is open.
    func closeSegment(reason: SegmentCloseReason) async throws
}

/// Receiver-side flags and storage hints carried in CHECKPOINT; provided by SegmentStore.
public protocol ReceiverPolicy: Sendable {
    /// Bitmask of ProtocolConstants RECEIVER_FLAG_* values.
    func receiverFlags() -> UInt32
    func freeStorageHintKb() -> UInt32
}

/// Persisted state that lets a restarted process resume a receiver session.
public struct ReceiverResumeState: Sendable, Equatable, Codable {
    public let lastStreamId: UInt32
    /// Highest contiguous sequence persisted, or nil when nothing was persisted.
    public let lastContiguousSequence: UInt32?
    public let lastSampleIndex: UInt64

    public init(lastStreamId: UInt32, lastContiguousSequence: UInt32?, lastSampleIndex: UInt64) {
        self.lastStreamId = lastStreamId
        self.lastContiguousSequence = lastContiguousSequence
        self.lastSampleIndex = lastSampleIndex
    }
}

public protocol ReceiverResumeStore: Sendable {
    func save(_ state: ReceiverResumeState) async
    func load() async -> ReceiverResumeState?
    func clear() async
}

File: core/transport/src/commonMain/kotlin/dev/audiocompanion/transport/SegmentSink.kt
package dev.audiocompanion.transport

import dev.audiocompanion.protocol.StreamStart

/** One encoded audio frame, addressed by stream sequence and cumulative sample index. */
class SegmentFrame(
    val sequence: UInt,
    val sampleIndex: ULong,
    val payload: ByteArray,
) {
    override fun toString(): String =
        "SegmentFrame(sequence=$sequence, sampleIndex=$sampleIndex, len=${payload.size})"
}

/** Why a gap record exists in a segment. */
sealed interface GapOrigin {
    /** The watch explicitly reported the gap via STREAM_GAP. */
    data class WatchReported(val reasonRaw: Int, val watchDropCounter: UInt) : GapOrigin

    /** Synthesized by the receiver after observing a sequence discontinuity in STREAM_DATA. */
    data object SequenceSkip : GapOrigin
}

data class GapRecord(
    val firstMissingSequence: UInt,
    /** 0 = unknown / elapsed-time-only gap. */
    val missingFrameCount: UInt,
    val firstMissingSampleIndex: ULong,
    val origin: GapOrigin,
)

/** Why a segment was closed. */
sealed interface SegmentCloseReason {
    /** The watch sent STREAM_STOP. */
    data class Stopped(
        val reasonRaw: Int,
        val finalSequence: UInt,
        val finalSampleIndex: ULong,
    ) : SegmentCloseReason

    /** The BLE link dropped (or the session ended) while the stream was open. */
    data object Interrupted : SegmentCloseReason

    /** A new STREAM_START arrived while a segment was still open. */
    data object Superseded : SegmentCloseReason
}

/** Diagnostics-only provenance attached to a segment. */
data class SegmentProvenance(
    val fwVersionPacked: UInt,
    val protocolVersion: Int,
)

/**
 * Durable storage seam consumed by [AudioReceiverSession]; implemented by :core:storage.
 * All functions are suspend and must only return once the data is durable — the session
 * computes checkpoint sequences from what these calls have accepted.
 */
interface SegmentSink {
    /** Opens a segment for a newly started stream. Any previously open segment was closed. */
    suspend fun openSegment(start: StreamStart, receivedAtMs: Long, provenance: SegmentProvenance?)

    /**
     * Appends frames to the open segment; durable on return. Returns the frames actually
     * persisted — a post-RESUME spool rewind re-sends frames the sink may already hold, and
     * downstream consumers (live waveform, live transcription) must only see what was new.
     */
    suspend fun appendFrames(streamId: UInt, frames: List<SegmentFrame>): List<SegmentFrame>

    /** Records a gap (watch-reported or synthesized) in the open segment. */
    suspend fun recordGap(streamId: UInt, gap: GapRecord)

    /** Closes the open segment. No-op when no segment is open. */
    suspend fun closeSegment(reason: SegmentCloseReason)
}

/** Receiver-side flags and storage hints carried in CHECKPOINT; provided by :core:storage. */
interface ReceiverPolicy {
    /** Bitmask of ProtocolConstants.RECEIVER_FLAG_*. */
    fun receiverFlags(): UInt

    fun freeStorageHintKb(): UInt
}

/** Persisted state that lets a restarted process resume a receiver session. */
data class ReceiverResumeState(
    val lastStreamId: UInt,
    /** Highest contiguous sequence persisted, or null when nothing was persisted. */
    val lastContiguousSequence: UInt?,
    val lastSampleIndex: ULong,
)

interface ReceiverResumeStore {
    suspend fun save(state: ReceiverResumeState)

    suspend fun load(): ReceiverResumeState?

    suspend fun clear() {}
}
